File: link/protocol/schema-lite.mjs
import { isUtcTimestamp } from "./time.mjs";
import { SCHEMA_DATA } from "./schema-data.mjs";

export function envelopeSchema() {
  return SCHEMA_DATA.envelope;
}

export function responseSchema() {
  return SCHEMA_DATA.response;
}

const PAYLOAD_KEYS = {
  health: "health",
  handshake: "handshake",
  "application.queryCandidates": "query-candidates",
  "job.save": "job-save",
  "fill.submit": "fill-submit",
  "snapshot.chunk": "snapshot-chunk",
  "submit.confirm": "submit-confirm",
  "outbox.reconcile": "outbox-reconcile",
};

const RESPONSE_PAYLOAD_KEYS = {
  health: "health",
  handshake: "handshake",
  "application.queryCandidates": "query-candidates",
  "job.save": "write",
  "fill.submit": "write",
  "submit.confirm": "write",
  "snapshot.chunk": "snapshot-chunk",
  "outbox.reconcile": "outbox-reconcile",
};

export function payloadSchema(messageType) {
  const key = PAYLOAD_KEYS[messageType];
  return key ? SCHEMA_DATA.payloads[key] : null;
}

export function responsePayloadSchema(messageType) {
  const key = RESPONSE_PAYLOAD_KEYS[messageType];
  return key ? SCHEMA_DATA.responses[key] : null;
}

export const RULES = SCHEMA_DATA.rules;

function fail(message) {
  const err = new Error(`invalid_payload: ${message}`);
  err.code = "invalid_payload";
  err.layer = "structure";
  err.retryable = false;
  return err;
}

function isUuid(value) {
  const parts = value.split("-");
  return (
    parts.length === 5 &&
    parts[0].length === 8 &&
    parts[1].length === 4 &&
    parts[2].length === 4 &&
    parts[3].length === 4 &&
    parts[4].length === 12 &&
    [...value].every((c) => /[0-9a-fA-F-]/.test(c))
  );
}

function syntacticTimestamp(value) {
  if (typeof value !== "string" || !value.endsWith("Z") || value.length < 20) return false;
  const m = value.match(/^([0-9]{4})-([0-9]{2})-([0-9]{2})T([0-9]{2}):([0-9]{2}):([0-9]{2})(\.[0-9]+)?Z$/);
  return Boolean(m);
}

function patternMatches(value, pattern) {
  if (pattern.includes("[0-9a-fA-F]{8}-")) return isUuid(value);
  if (pattern === "^[0-9a-f]{64}$") return /^[0-9a-f]{64}$/.test(value);
  if (pattern.includes("T[0-9]{2}:[0-9]{2}:[0-9]{2}")) return syntacticTimestamp(value);
  throw fail(`unsupported schema pattern ${pattern}`);
}

export function validateSchema(instance, schema) {
  if (schema.type === "object") {
    if (!instance || typeof instance !== "object" || Array.isArray(instance)) {
      throw fail("value must be an object");
    }
    const properties = schema.properties || {};
    if (schema.additionalProperties === false) {
      for (const key of Object.keys(instance)) {
        if (!(key in properties)) throw fail(`unexpected field ${key}`);
      }
    }
    for (const key of schema.required || []) {
      if (!Object.prototype.hasOwnProperty.call(instance, key)) {
        throw fail(`missing field ${key}`);
      }
    }
    for (const [key, sub] of Object.entries(properties)) {
      if (Object.prototype.hasOwnProperty.call(instance, key)) {
        validateSchema(instance[key], sub);
      }
    }
  } else if (schema.type === "array") {
    if (!Array.isArray(instance)) throw fail("value must be an array");
    if (schema.minItems != null && instance.length < schema.minItems) throw fail("array too short");
    if (schema.maxItems != null && instance.length > schema.maxItems) throw fail("array too long");
    if (schema.items) {
      for (const item of instance) validateSchema(item, schema.items);
    }
  } else if (schema.type === "string") {
    if (typeof instance !== "string") throw fail("value must be a string");
    if (schema.minLength != null && [...instance].length < schema.minLength) throw fail("string too short");
    if (schema.maxLength != null && [...instance].length > schema.maxLength) throw fail("string too long");
    if (schema.pattern && !patternMatches(instance, schema.pattern)) throw fail("string does not match pattern");
  } else if (schema.type === "integer") {
    if (!Number.isInteger(instance)) throw fail("value must be an integer");
    if (schema.minimum != null && instance < schema.minimum) throw fail("integer below minimum");
    if (schema.maximum != null && instance > schema.maximum) throw fail("integer above maximum");
  } else if (schema.type === "boolean") {
    if (typeof instance !== "boolean") throw fail("value must be a boolean");
  }
  if (schema.enum && !schema.enum.some((v) => Object.is(v, instance) || v === instance)) {
    throw fail("value is not in enum");
  }
}

export { isUtcTimestamp };

File: desktop/crates/protocol/src/schema_lite.rs
use serde_json::Value;

use crate::error::{ErrorCode, Layer, ProtocolError};

pub fn envelope_schema() -> Value {
    serde_json::from_str(include_str!("../schemas/request-envelope.json")).expect("envelope schema")
}

pub fn response_schema() -> Value {
    serde_json::from_str(include_str!("../schemas/response-envelope.json")).expect("response schema")
}

pub fn payload_schema(message_type: &str) -> Option<Value> {
    let raw = match message_type {
        "health" => include_str!("../schemas/payloads/health.json"),
        "handshake" => include_str!("../schemas/payloads/handshake.json"),
        "application.queryCandidates" => include_str!("../schemas/payloads/query-candidates.json"),
        "job.save" => include_str!("../schemas/payloads/job-save.json"),
        "fill.submit" => include_str!("../schemas/payloads/fill-submit.json"),
        "snapshot.chunk" => include_str!("../schemas/payloads/snapshot-chunk.json"),
        "submit.confirm" => include_str!("../schemas/payloads/submit-confirm.json"),
        "outbox.reconcile" => include_str!("../schemas/payloads/outbox-reconcile.json"),
        _ => return None,
    };
    Some(serde_json::from_str(raw).expect("payload schema"))
}

pub fn response_payload_schema(message_type: &str) -> Option<Value> {
    let raw = match message_type {
        "health" => include_str!("../schemas/responses/health.json"),
        "handshake" => include_str!("../schemas/responses/handshake.json"),
        "application.queryCandidates" => include_str!("../schemas/responses/query-candidates.json"),
        "job.save" | "fill.submit" | "submit.confirm" => include_str!("../schemas/responses/write.json"),
        "snapshot.chunk" => include_str!("../schemas/responses/snapshot-chunk.json"),
        "outbox.reconcile" => include_str!("../schemas/responses/outbox-reconcile.json"),
        _ => return None,
    };
    Some(serde_json::from_str(raw).expect("response payload schema"))
}

pub fn validate_schema(instance: &Value, schema: &Value) -> Result<(), ProtocolError> {
    apply(instance, schema)
}

fn apply(instance: &Value, schema: &Value) -> Result<(), ProtocolError> {
    let schema_obj = schema.as_object().ok_or_else(|| invalid("schema must be an object"))?;
    if let Some(ty) = schema_obj.get("type").and_then(Value::as_str) {
        match ty {
            "object" => {
                let obj = instance.as_object().ok_or_else(|| invalid("value must be an object"))?;
                if instance.as_array().is_some() {
                    return Err(invalid("value must be an object"));
                }
                if schema_obj.get("additionalProperties") == Some(&Value::Bool(false)) {
                    let allowed = schema_obj
                        .get("properties")
                        .and_then(Value::as_object)
                        .map(|p| p.keys().cloned().collect::<Vec<_>>())
                        .unwrap_or_default();
                    for key in obj.keys() {
                        if !allowed.iter().any(|a| a == key) {
                            return Err(invalid(&format!("unexpected field {key}")));
                        }
                    }
                }
                if let Some(required) = schema_obj.get("required").and_then(Value::as_array) {
                    for key in required {
                        let name = key.as_str().unwrap_or("");
                        if !obj.contains_key(name) {
                            return Err(invalid(&format!("missing field {name}")));
                        }
                    }
                }
                if let Some(props) = schema_obj.get("properties").and_then(Value::as_object) {
                    for (key, sub) in props {
                        if let Some(value) = obj.get(key) {
                            apply(value, sub)?;
                        }
                    }
                }
            }
            "array" => {
                let arr = instance.as_array().ok_or_else(|| invalid("value must be an array"))?;
                if let Some(min) = schema_obj.get("minItems").and_then(Value::as_u64) {
                    if (arr.len() as u64) < min {
                        return Err(invalid("array too short"));
                    }
                }
                if let Some(max) = schema_obj.get("maxItems").and_then(Value::as_u64) {
                    if (arr.len() as u64) > max {
                        return Err(invalid("array too long"));
                    }
                }
                if let Some(items) = schema_obj.get("items") {
                    for item in arr {
                        apply(item, items)?;
                    }
                }
            }
            "string" => {
                let s = instance.as_str().ok_or_else(|| invalid("value must be a string"))?;
                if let Some(min) = schema_obj.get("minLength").and_then(Value::as_u64) {
                    if (s.chars().count() as u64) < min {
                        return Err(invalid("string too short"));
                    }
                }
                if let Some(max) = schema_obj.get("maxLength").and_then(Value::as_u64) {
                    if (s.chars().count() as u64) > max {
                        return Err(invalid("string too long"));
                    }
                }
                if let Some(pattern) = schema_obj.get("pattern").and_then(Value::as_str) {
                    match check_pattern(s, pattern) {
                        PatternCheck::Match => {}
                        PatternCheck::Mismatch => return Err(invalid("string does not match pattern")),
                        PatternCheck::Unsupported => {
                            return Err(invalid(&format!("unsupported schema pattern {pattern}")));
                        }
                    }
                }
            }
            "integer" => {
                let n = instance.as_i64().ok_or_else(|| invalid("value must be an integer"))?;
                if let Some(min) = schema_obj.get("minimum").and_then(Value::as_i64) {
                    if n < min {
                        return Err(invalid("integer below minimum"));
                    }
                }
                if let Some(max) = schema_obj.get("maximum").and_then(Value::as_i64) {
                    if n > max {
                        return Err(invalid("integer above maximum"));
                    }
                }
            }
            "boolean" => {
                if !instance.is_boolean() {
                    return Err(invalid("value must be a boolean"));
                }
            }
            other => return Err(invalid(&format!("unsupported schema type {other}"))),
        }
    }
    if let Some(values) = schema_obj.get("enum").and_then(Value::as_array) {
        if !values.iter().any(|v| v == instance) {
            return Err(invalid("value is not in enum"));
        }
    }
    Ok(())
}

fn invalid(message: &str) -> ProtocolError {
    ProtocolError::new(ErrorCode::InvalidPayload, Layer::Structure, message)
}

enum PatternCheck {
    Match,
    Mismatch,
    Unsupported,
}

fn check_pattern(value: &str, pattern: &str) -> PatternCheck {
    if pattern.contains("[0-9a-fA-F]{8}-") {
        return if is_uuid(value) {
            PatternCheck::Match
        } else {
            PatternCheck::Mismatch
        };
    }
    if pattern == "^[0-9a-f]{64}$" {
        return if value.len() == 64 && value.chars().all(|c| matches!(c, '0'..='9' | 'a'..='f')) {
            PatternCheck::Match
        } else {
            PatternCheck::Mismatch
        };
    }
    if pattern.contains("T[0-9]{2}:[0-9]{2}:[0-9]{2}") {
        return if timestamp_matches_schema_pattern(value) {
            PatternCheck::Match
        } else {
            PatternCheck::Mismatch
        };
    }
    PatternCheck::Unsupported
}

fn syntactic_timestamp(value: &str) -> bool {
    let b = value.as_bytes();
    if !value.ends_with('Z') || !value.is_ascii() || b.len() < 20 {
        return false;
    }
    let digits = |range: std::ops::Range<usize>| range.clone().all(|i| b.get(i).copied().map(|c| c.is_ascii_digit()).unwrap_or(false));
    digits(0..4)
        && b.get(4) == Some(&b'-')
        && digits(5..7)
        && b.get(7) == Some(&b'-')
        && digits(8..10)
        && b.get(10) == Some(&b'T')
        && digits(11..13)
        && b.get(13) == Some(&b':')
        && digits(14..16)
        && b.get(16) == Some(&b':')
        && digits(17..19)
        && (b.len() == 20 || (b.len() > 21 && b[19] == b'.' && b[20..b.len() - 1].iter().all(u8::is_ascii_digit)))
}

fn is_uuid(value: &str) -> bool {
    let parts: Vec<&str> = value.split('-').collect();
    parts.len() == 5
        && parts[0].len() == 8
        && parts[1].len() == 4
        && parts[2].len() == 4
        && parts[3].len() == 4
        && parts[4].len() == 12
        && value.chars().all(|c| c.is_ascii_hexdigit() || c == '-')
}

/// Schema pattern only. Calendar checking is a separate protocol rule.
pub fn timestamp_matches_schema_pattern(value: &str) -> bool {
    syntactic_timestamp(value)
}
